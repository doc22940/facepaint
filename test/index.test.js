/* eslint-disable no-sparse-arrays */
import React from 'react'
import renderer from 'react-test-renderer'
import { css, sheet, flush, cx, registered } from 'emotion'

import facepaint from '../src/index'

const mq = facepaint(
  [
    '@media(min-width: 420px)',
    '@media(min-width: 920px)',
    '@media(min-width: 1120px)',
    '@media(min-width: 11200px)'
  ],
  { overlap: true }
)

const mql = facepaint(
  [
    '@media(min-width: 420px)',
    '@media(min-width: 920px)',
    '@media(min-width: 1120px)',
    '@media(min-width: 11200px)'
  ],
  { literal: true }
)

const pseudo = facepaint([':hover', ':active', ':focus'])

describe('facepaint', () => {
  afterEach(() => flush())
  test('basic', () => {
    const result = css(mq({ color: ['red', 'green', 'blue', 'darkorchid'] }))
    const tree = renderer.create(<div css={result}>Basic</div>).toJSON()
    expect(tree).toMatchSnapshot()
    expect(sheet).toMatchSnapshot()
  })

  test('holes', () => {
    const result = css(mq({ color: ['red', , 'blue', 'darkorchid'] }))
    const tree = renderer.create(<div css={result}>Basic</div>).toJSON()
    expect(tree).toMatchSnapshot()
    expect(sheet).toMatchSnapshot()
  })

  test('undefined', () => {
    const result = css(mq({ color: ['red', undefined, 'blue', 'darkorchid'] }))
    const tree = renderer.create(<div css={result}>Basic</div>).toJSON()
    expect(tree).toMatchSnapshot()
    expect(sheet).toMatchSnapshot()
  })

  test('repeating', () => {
    const result = css(
      mq({ color: ['red', 'blue', undefined, 'blue', 'darkorchid'] })
    )
    const tree = renderer.create(<div css={result}>Basic</div>).toJSON()
    expect(tree).toMatchSnapshot()
    expect(sheet).toMatchSnapshot()
  })

  test('nested arrays', () => {
    const result = css(mq([[[[{ color: ['red', 'blue', 'darkorchid'] }]]]]))
    const tree = renderer.create(<div css={result}>Basic</div>).toJSON()
    expect(tree).toMatchSnapshot()
    expect(sheet).toMatchSnapshot()
  })

  test('multiple', () => {
    const result = css(
      mq({
        color: ['red', 'green', 'blue', 'darkorchid'],
        display: ['flex', 'block', 'inline-block', 'table'],
        fontSize: 12,
        alignItems: 'center'
      })
    )
    const tree = renderer.create(<div css={result}>multiple</div>).toJSON()
    expect(tree).toMatchSnapshot()
    expect(sheet).toMatchSnapshot()
  })

  test('nested', () => {
    const result = css(
      mq({
        backgroundColor: 'hotpink',
        textAlign: 'center',
        width: ['25%', '50%', '75%', '100%'],
        '& .foo': {
          color: ['red', 'green', 'blue', 'darkorchid'],
          '& img': {
            height: [10, 15, 20, 25]
          }
        }
      })
    )
    const tree = renderer
      .create(
        <div css={result}>
          <div className="foo">foo</div>
          function
        </div>
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
    expect(sheet).toMatchSnapshot()
  })

  test('pseudo', () => {
    const result = css(
      pseudo({
        backgroundColor: 'hotpink',
        textAlign: 'center',
        width: ['25%', '50%', '75%', '100%'],
        '& .foo': {
          color: ['red', 'green', 'blue', 'darkorchid'],
          '& img': {
            height: [10, 15, 20, 25]
          }
        }
      })
    )
    const tree = renderer
      .create(
        <div css={result}>
          <div className="foo">foo</div>
          function
        </div>
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
    expect(sheet).toMatchSnapshot()
  })

  test('array values with selectors', () => {
    const result = css(
      mq({
        '& .current-index': [
          {
            color: ['blue', 'red']
          },
          {
            marginRight: 15,
            display: ['none', 'block'],
            letterSpacing: 3
          }
        ]
      })
    )
    const tree = renderer
      .create(
        <div css={result}>
          <div className="foo">foo</div>
          function
        </div>
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
    expect(sheet).toMatchSnapshot()
  })

  test('boolean, null, and undefined values', () => {
    const result = css(
      mq(
        { color: 'blue' },
        1 === 2 && { color: 'green' },
        false,
        true,
        undefined,
        null,
        [
          { color: 'red' },
          1 === 2 && { color: 'green' },
          false,
          true,
          undefined,
          null
        ]
      )
    )

    const tree = renderer
      .create(
        <div css={result}>
          <div className="foo">foo</div>
          function
        </div>
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
    expect(sheet).toMatchSnapshot()
  })

  test('composition', () => {
    const a = css(mq({ background: ['green', 'blue'] }))
    const b = css(mq({ background: 'orange' }))
    const c = css(mq({ background: ['orange', 'orange'] }))
    const d = css(mq({ background: ['orange', 'orange', 'orange', 'orange'] }))

    const tree = renderer.create(<div css={cx(a, b, c, d)} />).toJSON()
    expect(tree).toMatchSnapshot()
    expect(sheet).toMatchSnapshot()
  })

  test('more composition', () => {
    const styles1 = css(mq({ marginTop: [1, 2] }))
    const styles2 = css(mq({ marginTop: [500, 500] }))
    const tree = renderer.create(<div css={cx(styles1, styles2)} />).toJSON()
    expect(tree).toMatchSnapshot()
    expect(sheet).toMatchSnapshot()
  })

  test('literal 1', () => {
    const result = css(mql({ background: ['red'] }))
    expect(result).toMatchSnapshot()
    const tree = renderer.create(<div css={result}>foo</div>).toJSON()

    expect(tree).toMatchSnapshot()
    expect(sheet).toMatchSnapshot()
  })

  test('literal 2', () => {
    const result = css(mql({ background: ['red', 'green'] }))
    expect(result).toMatchSnapshot()
    const tree = renderer.create(<div css={result}>foo</div>).toJSON()

    expect(tree).toMatchSnapshot()
    expect(sheet).toMatchSnapshot()
  })

  test('literal all', () => {
    const result = css(mql({ background: ['red', 'green', 'blue', 'orange'] }))
    expect(result).toMatchSnapshot()
    const tree = renderer.create(<div css={result}>foo</div>).toJSON()

    expect(tree).toMatchSnapshot()
    expect(sheet).toMatchSnapshot()
  })

  test('literal: prevent unexpected selector', () => {
    const styles1 = css(mql({ marginTop: [1, 2] }))
    const styles2 = css(mql({ marginTop: [500, 500] }))
    const result = cx(styles1, styles2)
    expect(result).toMatchSnapshot()
    const tree = renderer.create(<div css={result}>foo</div>).toJSON()

    expect(tree).toMatchSnapshot()
    expect(sheet).toMatchSnapshot()
  })
  test('https://github.com/emotion-js/facepaint/issues/9', () => {
    const lolol = facepaint([
      '@media (min-width: 400px)',
      '@media (min-width: 700px)',
      '@media print'
    ])
    css(
      lolol({
        background: ['white'],
        borderBottom: ['1px solid black', null, 0],
        height: ['360px', '350px'],
        padding: ['8px 15px 3px', '11px 15px 6px', '4px 15px 2px']
      })
    )
    expect(sheet).toMatchSnapshot()
  })
})
