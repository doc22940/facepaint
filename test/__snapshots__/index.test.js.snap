// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`facepaint array values with selectors 1`] = `
.glamor-0 .current-index {
  color: blue;
  margin-right: 15px;
  display: none;
  -webkit-letter-spacing: 3px;
  -moz-letter-spacing: 3px;
  -ms-letter-spacing: 3px;
  letter-spacing: 3px;
}

@media (min-width:420px) {
  .glamor-0 .current-index {
    color: red;
  }
}

@media (min-width:420px) {
  .glamor-0 .current-index {
    display: block;
  }
}

<div
  className="glamor-0"
>
  <div
    className="foo"
  >
    foo
  </div>
  function
</div>
`;

exports[`facepaint array values with selectors 2`] = `
".css-1ot4erq .current-index {
  color: blue;
  margin-right: 15px;
  display: none;
  -webkit-letter-spacing: 3px;
  -moz-letter-spacing: 3px;
  -ms-letter-spacing: 3px;
  letter-spacing: 3px;
}

@media (min-width:420px) {
  .css-1ot4erq .current-index {
    color: red;
  }
}

@media (min-width:420px) {
  .css-1ot4erq .current-index {
    display: block;
  }
}"
`;

exports[`facepaint basic 1`] = `
.glamor-0 {
  color: red;
}

@media (min-width:420px) {
  .glamor-0 {
    color: green;
  }
}

@media (min-width:920px) {
  .glamor-0 {
    color: blue;
  }
}

@media (min-width:1120px) {
  .glamor-0 {
    color: darkorchid;
  }
}

<div
  className="glamor-0"
>
  Basic
</div>
`;

exports[`facepaint basic 2`] = `
".css-xee1ha {
  color: red;
}

@media (min-width:420px) {
  .css-xee1ha {
    color: green;
  }
}

@media (min-width:920px) {
  .css-xee1ha {
    color: blue;
  }
}

@media (min-width:1120px) {
  .css-xee1ha {
    color: darkorchid;
  }
}"
`;

exports[`facepaint boolean, null, and undefined values 1`] = `
.glamor-0 {
  color: blue;
  color: red;
}

<div
  className="glamor-0"
>
  <div
    className="foo"
  >
    foo
  </div>
  function
</div>
`;

exports[`facepaint boolean, null, and undefined values 2`] = `
".css-dhb7kq {
  color: blue;
  color: red;
}"
`;

exports[`facepaint composition 1`] = `
.glamor-0 {
  background: green;
  background: orange;
  background: orange;
  background: orange;
}

@media (min-width:420px) {
  .glamor-0 {
    background: blue;
  }
}

<div
  className="glamor-0"
/>
`;

exports[`facepaint composition 2`] = `
".css-16gx40p {
  background: green;
}

@media (min-width:420px) {
  .css-16gx40p {
    background: blue;
  }
}

.css-1y2rzxt {
  background: orange;
}

.css-1plng0y {
  background: green;
  background: orange;
  background: orange;
  background: orange;
}

@media (min-width:420px) {
  .css-1plng0y {
    background: blue;
  }
}"
`;

exports[`facepaint holes 1`] = `
.glamor-0 {
  color: red;
}

@media (min-width:920px) {
  .glamor-0 {
    color: blue;
  }
}

@media (min-width:1120px) {
  .glamor-0 {
    color: darkorchid;
  }
}

<div
  className="glamor-0"
>
  Basic
</div>
`;

exports[`facepaint holes 2`] = `
".css-196uzh {
  color: red;
}

@media (min-width:920px) {
  .css-196uzh {
    color: blue;
  }
}

@media (min-width:1120px) {
  .css-196uzh {
    color: darkorchid;
  }
}"
`;

exports[`facepaint https://github.com/emotion-js/facepaint/issues/9 1`] = `
".css-1eyh3wc {
  background: white;
  border-bottom: 1px solid black;
  height: 360px;
  padding: 8px 15px 3px;
}

@media (min-width:400px) {
  .css-1eyh3wc {
    height: 350px;
    padding: 11px 15px 6px;
  }
}

@media (min-width:700px) {
  .css-1eyh3wc {
    border-bottom: 0;
    padding: 4px 15px 2px;
  }
}"
`;

exports[`facepaint literal 1 1`] = `"css-1n438h9"`;

exports[`facepaint literal 1 2`] = `
@media (min-width:420px) {
  .glamor-0 {
    background: red;
  }
}

<div
  className="glamor-0"
>
  foo
</div>
`;

exports[`facepaint literal 1 3`] = `
"@media (min-width:420px) {
  .css-1n438h9 {
    background: red;
  }
}"
`;

exports[`facepaint literal 2 1`] = `"css-1k7f7mv"`;

exports[`facepaint literal 2 2`] = `
@media (min-width:420px) {
  .glamor-0 {
    background: red;
  }
}

@media (min-width:920px) {
  .glamor-0 {
    background: green;
  }
}

<div
  className="glamor-0"
>
  foo
</div>
`;

exports[`facepaint literal 2 3`] = `
"@media (min-width:420px) {
  .css-1k7f7mv {
    background: red;
  }
}

@media (min-width:920px) {
  .css-1k7f7mv {
    background: green;
  }
}"
`;

exports[`facepaint literal all 1`] = `"css-1ahnw43"`;

exports[`facepaint literal all 2`] = `
@media (min-width:420px) {
  .glamor-0 {
    background: red;
  }
}

@media (min-width:920px) {
  .glamor-0 {
    background: green;
  }
}

@media (min-width:1120px) {
  .glamor-0 {
    background: blue;
  }
}

@media (min-width:11200px) {
  .glamor-0 {
    background: orange;
  }
}

<div
  className="glamor-0"
>
  foo
</div>
`;

exports[`facepaint literal all 3`] = `
"@media (min-width:420px) {
  .css-1ahnw43 {
    background: red;
  }
}

@media (min-width:920px) {
  .css-1ahnw43 {
    background: green;
  }
}

@media (min-width:1120px) {
  .css-1ahnw43 {
    background: blue;
  }
}

@media (min-width:11200px) {
  .css-1ahnw43 {
    background: orange;
  }
}"
`;

exports[`facepaint literal: prevent unexpected selector 1`] = `"css-15ynynd"`;

exports[`facepaint literal: prevent unexpected selector 2`] = `
@media (min-width:420px) {
  .glamor-0 {
    margin-top: 1px;
  }
}

@media (min-width:920px) {
  .glamor-0 {
    margin-top: 2px;
  }
}

@media (min-width:420px) {
  .glamor-0 {
    margin-top: 500px;
  }
}

@media (min-width:920px) {
  .glamor-0 {
    margin-top: 500px;
  }
}

<div
  className="glamor-0"
>
  foo
</div>
`;

exports[`facepaint literal: prevent unexpected selector 3`] = `
"@media (min-width:420px) {
  .css-1ca8ec {
    margin-top: 1px;
  }
}

@media (min-width:920px) {
  .css-1ca8ec {
    margin-top: 2px;
  }
}

@media (min-width:420px) {
  .css-13rmv1r {
    margin-top: 500px;
  }
}

@media (min-width:920px) {
  .css-13rmv1r {
    margin-top: 500px;
  }
}

@media (min-width:420px) {
  .css-15ynynd {
    margin-top: 1px;
  }
}

@media (min-width:920px) {
  .css-15ynynd {
    margin-top: 2px;
  }
}

@media (min-width:420px) {
  .css-15ynynd {
    margin-top: 500px;
  }
}

@media (min-width:920px) {
  .css-15ynynd {
    margin-top: 500px;
  }
}"
`;

exports[`facepaint more composition 1`] = `
.glamor-0 {
  margin-top: 1px;
  margin-top: 500px;
}

@media (min-width:420px) {
  .glamor-0 {
    margin-top: 2px;
  }
}

<div
  className="glamor-0"
/>
`;

exports[`facepaint more composition 2`] = `
".css-1lg6icu {
  margin-top: 1px;
}

@media (min-width:420px) {
  .css-1lg6icu {
    margin-top: 2px;
  }
}

.css-lk2z8m {
  margin-top: 500px;
}

.css-1874a9x {
  margin-top: 1px;
  margin-top: 500px;
}

@media (min-width:420px) {
  .css-1874a9x {
    margin-top: 2px;
  }
}"
`;

exports[`facepaint multiple 1`] = `
.glamor-0 {
  color: red;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  font-size: 12px;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

@media (min-width:420px) {
  .glamor-0 {
    color: green;
    display: block;
  }
}

@media (min-width:920px) {
  .glamor-0 {
    color: blue;
    display: inline-block;
  }
}

@media (min-width:1120px) {
  .glamor-0 {
    color: darkorchid;
    display: table;
  }
}

<div
  className="glamor-0"
>
  multiple
</div>
`;

exports[`facepaint multiple 2`] = `
".css-egce6e {
  color: red;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  font-size: 12px;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

@media (min-width:420px) {
  .css-egce6e {
    color: green;
    display: block;
  }
}

@media (min-width:920px) {
  .css-egce6e {
    color: blue;
    display: inline-block;
  }
}

@media (min-width:1120px) {
  .css-egce6e {
    color: darkorchid;
    display: table;
  }
}"
`;

exports[`facepaint nested 1`] = `
.glamor-0 {
  background-color: hotpink;
  text-align: center;
  width: 25%;
}

.glamor-0 .foo {
  color: red;
}

.glamor-0 .foo img {
  height: 10px;
}

@media (min-width:420px) {
  .glamor-0 {
    width: 50%;
  }
}

@media (min-width:920px) {
  .glamor-0 {
    width: 75%;
  }
}

@media (min-width:1120px) {
  .glamor-0 {
    width: 100%;
  }
}

@media (min-width:420px) {
  .glamor-0 .foo {
    color: green;
  }
}

@media (min-width:920px) {
  .glamor-0 .foo {
    color: blue;
  }
}

@media (min-width:1120px) {
  .glamor-0 .foo {
    color: darkorchid;
  }
}

@media (min-width:420px) {
  .glamor-0 .foo img {
    height: 15px;
  }
}

@media (min-width:920px) {
  .glamor-0 .foo img {
    height: 20px;
  }
}

@media (min-width:1120px) {
  .glamor-0 .foo img {
    height: 25px;
  }
}

<div
  className="glamor-0"
>
  <div
    className="foo"
  >
    foo
  </div>
  function
</div>
`;

exports[`facepaint nested 2`] = `
".css-rbuh8g {
  background-color: hotpink;
  text-align: center;
  width: 25%;
}

@media (min-width:420px) {
  .css-rbuh8g {
    width: 50%;
  }
}

@media (min-width:920px) {
  .css-rbuh8g {
    width: 75%;
  }
}

@media (min-width:1120px) {
  .css-rbuh8g {
    width: 100%;
  }
}

.css-rbuh8g .foo {
  color: red;
}

@media (min-width:420px) {
  .css-rbuh8g .foo {
    color: green;
  }
}

@media (min-width:920px) {
  .css-rbuh8g .foo {
    color: blue;
  }
}

@media (min-width:1120px) {
  .css-rbuh8g .foo {
    color: darkorchid;
  }
}

.css-rbuh8g .foo img {
  height: 10px;
}

@media (min-width:420px) {
  .css-rbuh8g .foo img {
    height: 15px;
  }
}

@media (min-width:920px) {
  .css-rbuh8g .foo img {
    height: 20px;
  }
}

@media (min-width:1120px) {
  .css-rbuh8g .foo img {
    height: 25px;
  }
}"
`;

exports[`facepaint nested arrays 1`] = `
.glamor-0 {
  color: red;
}

@media (min-width:420px) {
  .glamor-0 {
    color: blue;
  }
}

@media (min-width:920px) {
  .glamor-0 {
    color: darkorchid;
  }
}

<div
  className="glamor-0"
>
  Basic
</div>
`;

exports[`facepaint nested arrays 2`] = `
".css-1faqh3h {
  color: red;
}

@media (min-width:420px) {
  .css-1faqh3h {
    color: blue;
  }
}

@media (min-width:920px) {
  .css-1faqh3h {
    color: darkorchid;
  }
}"
`;

exports[`facepaint pseudo 1`] = `
.glamor-0 {
  background-color: hotpink;
  text-align: center;
  width: 25%;
}

.glamor-0:hover {
  width: 50%;
}

.glamor-0:active {
  width: 75%;
}

.glamor-0:focus {
  width: 100%;
}

.glamor-0 .foo {
  color: red;
}

.glamor-0 .foo:hover {
  color: green;
}

.glamor-0 .foo:active {
  color: blue;
}

.glamor-0 .foo:focus {
  color: darkorchid;
}

.glamor-0 .foo img {
  height: 10px;
}

.glamor-0 .foo img:hover {
  height: 15px;
}

.glamor-0 .foo img:active {
  height: 20px;
}

.glamor-0 .foo img:focus {
  height: 25px;
}

<div
  className="glamor-0"
>
  <div
    className="foo"
  >
    foo
  </div>
  function
</div>
`;

exports[`facepaint pseudo 2`] = `
".css-1guvnfu {
  background-color: hotpink;
  text-align: center;
  width: 25%;
}

.css-1guvnfu:hover {
  width: 50%;
}

.css-1guvnfu:active {
  width: 75%;
}

.css-1guvnfu:focus {
  width: 100%;
}

.css-1guvnfu .foo {
  color: red;
}

.css-1guvnfu .foo:hover {
  color: green;
}

.css-1guvnfu .foo:active {
  color: blue;
}

.css-1guvnfu .foo:focus {
  color: darkorchid;
}

.css-1guvnfu .foo img {
  height: 10px;
}

.css-1guvnfu .foo img:hover {
  height: 15px;
}

.css-1guvnfu .foo img:active {
  height: 20px;
}

.css-1guvnfu .foo img:focus {
  height: 25px;
}"
`;

exports[`facepaint repeating 1`] = `
.glamor-0 {
  color: red;
}

@media (min-width:420px) {
  .glamor-0 {
    color: blue;
  }
}

@media (min-width:11200px) {
  .glamor-0 {
    color: darkorchid;
  }
}

<div
  className="glamor-0"
>
  Basic
</div>
`;

exports[`facepaint repeating 2`] = `
".css-nhhz1q {
  color: red;
}

@media (min-width:420px) {
  .css-nhhz1q {
    color: blue;
  }
}

@media (min-width:11200px) {
  .css-nhhz1q {
    color: darkorchid;
  }
}"
`;

exports[`facepaint undefined 1`] = `
.glamor-0 {
  color: red;
}

@media (min-width:920px) {
  .glamor-0 {
    color: blue;
  }
}

@media (min-width:1120px) {
  .glamor-0 {
    color: darkorchid;
  }
}

<div
  className="glamor-0"
>
  Basic
</div>
`;

exports[`facepaint undefined 2`] = `
".css-196uzh {
  color: red;
}

@media (min-width:920px) {
  .css-196uzh {
    color: blue;
  }
}

@media (min-width:1120px) {
  .css-196uzh {
    color: darkorchid;
  }
}"
`;
